subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/ModuleMetadataParser.java

```diff
@@ -140,6 +141,8 @@ private void consumeVariant(JsonReader reader, MutableComponentVariantResolveMet
                 files = consumeFiles(reader);
             } else if (name.equals("dependencies")) {
                 dependencies = consumeDependencies(reader);
+            } else if (name.equals("dependencyConstraints")) {
```

Comment:
I'm not sure I like the fact that we are treating `dependencies` and `dependencyConstraints` quite separately in the module metadata and in the `ComponentVariant` model. It feels like these are slight variations on the same thing:
like 'strong' and 'weak' dependencies. I see that we don't support excludes for constraints, but I still think that it might be simpler if we consider this a single collection of 'dependencies'. I don't feel strongly about it, though, just a hunch.

Reply:
I think conceptually want them to be different. Although they share a common base. But I think in the future they will most likely diverge even more. For example, we want to support constraints for groups of dependencies that allow you to omit the _module name_. Something that is not possible for dependencies.

So I think it is good to separate the two concepts on the API side (both DSL and metadata). Although there is a bit of duplication now. 

On the DSL side, they still share a common super type (`Dependency`). But the reason for this is to allow them to be treated as one set of things when inspecting/manipulating `Configuration`s. Which is mostly driven by backward compatibility. In the future we plan to move further away from the `Configuration` towards the `Variant` API. We might then also treat them more separated there at some point.

Internally in dependency resolution, both are more similar. A dependency constraint is "just" a dependency marked as "pending". But that will also change more when we add things like support for _constraints for groups_.